internal/injector/injector.go

```diff
@@ -0,0 +1,464 @@
+			inject(name, value)
+
+			continue
+		}
+
+		// decrypts value with Vault Transit Secret Engine
+		if i.client.Transit.IsEncrypted(value) {
+			if len(i.config.TransitKeyID) == 0 {
+				return errors.Errorf("found encrypted variable, but transit key ID is empty: %s", name)
+			}
+
+			i.mu.RLock()
+			v, ok := i.transitCache[value]
+			i.mu.RUnlock()
+			if ok {
+				inject(name, string(v))
+
+				continue
+			}
+
+			out, err := i.client.Transit.Decrypt(i.config.TransitPath, i.config.TransitKeyID, []byte(value))
+			if err != nil {
+				if !i.config.IgnoreMissingSecrets {
+					return errors.Wrapf(err, "failed to decrypt variable: %s", name)
+				}
+				i.logger.Errorln("failed to decrypt variable:", name, err)
+
+				continue
+			}
+
+			i.mu.Lock()
+			i.transitCache[value] = out
+			i.mu.Unlock()
+
+			inject(name, string(out))
+
+			continue
+		}
+
+		split := strings.SplitN(valuePath, "#", 3)
+		valuePath = split[0]
+
+		if len(split) < 2 {
+			return errors.New("secret data key or template not defined")
+		}
+
+		key := split[1]
+
+		versionOrData := "-1"
+		if update {
+			versionOrData = "{}"
+		}
+		if len(split) == 3 {
+			versionOrData = split[2]
+		}
+
+		secretCacheKey := valuePath + "#" + versionOrData
+		var data map[string]interface{}
+		var err error
+
+		i.mu.RLock()
+		if data = i.secretCache[secretCacheKey]; data == nil {
+			data, err = i.readVaultPath(valuePath, versionOrData, update)
+		}
+		i.mu.RUnlock()
+
+		if err != nil {
+			return err
+		}
+
+		if data == nil {
+			if !i.config.IgnoreMissingSecrets {
+				return errors.Errorf("path not found: %s", valuePath)
+			}
+
+			i.logger.Warnf(
+				"Path not found: %s - We couldn't find a secret path. This is not an error since missing secrets can be ignored according to the configuration you've set (annotation: %s).",
+				valuePath,
+				MissingSecretsAnnotation,
+			)
+
+			continue
+		}
+
+		i.mu.Lock()
+		i.secretCache[secretCacheKey] = data
+		i.mu.Unlock()
+
+		templater := configuration.NewTemplater(configuration.DefaultLeftDelimiter, configuration.DefaultRightDelimiter)
+
+		if templater.IsGoTemplate(key) {
+			value, err := templater.Template(key, data)
+			if err != nil {
+				return errors.Wrapf(err, "failed to interpolate template key with vault data: %s", key)
+			}
+			inject(name, value.String())
+		} else {
+			if value, ok := data[key]; ok {
+				value, err := cast.ToStringE(value)
+				if err != nil {
+					return errors.Wrap(err, "value can't be cast to a string")
+				}
+				inject(name, value)
+			} else {
+				return errors.Errorf("key '%s' not found under path: %s", key, valuePath)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (i *SecretInjector) InjectSecretsFromVaultPath(paths string, inject SecretInjectorFunc) error {
+	vaultPaths := strings.Split(paths, ",")
+
+	for _, path := range vaultPaths {
+		split := strings.SplitN(path, "#", 2)
+		valuePath := split[0]
+
+		version := "-1"
+
+		if len(split) > 2 {
+			version = split[2]
+		}
```

Comment:
This will never be true since `len(split)` is always less (or equal) to 2 (check SplitN)

Reply:
My thinking was not to touch the contents of `internal/injector` and `internal/configuration` I copied from other projects until we figure out how to handle these project-wide internal packages, but if we decide we won't manage them in a central place, I'm going to change them drastically here 🙂 
That is a nice catch nevertheless, also reminded me to implement that if a secret version is specified, the workload shouldn't be reloaded based on only that even if that secret has a new version.

Reply:
That makes sense, @akijakya. Feel free to mark resolved the conversations that tackle copied internal packages which do not have logical issues.